spaghetti/clustering/GaussianMixtureClustering.py

```diff
@@ -0,0 +1,47 @@
+"""
+Copyright 2023 Spaghetti team
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+language governing permissions and limitations under the License.
+"""
+import numpy as np
+from sklearn.mixture import GaussianMixture
+from spaghetti.clustering.ClusteringModel import ClusteringModel
+
+class GaussianMixtureClustering(ClusteringModel):
+    """
+    Gaussian Mixture Clustering model. It is a subclass of the ClusteringModel class.
+    It uses the GaussianMixture model from sklearn for clustering data.
+    Example usages:
+        For fitting the model:
+            gmc = GaussianMixtureClustering()
+            gmc.fit(data)
+        For predicting clusters:
+            gmc = GaussianMixtureClustering()
+            clusters = gmc.predict(data)
+        For fitting the model and predicting clusters:
+            gmc = GaussianMixtureClustering()
+            clusters = gmc.fit_predict(data)
+
+    :param data: Data to be clustered. It should be a numpy array or similar data structure.
+    """
+    def __init__(self):
+        super().__init__()
```

Comment:
You should remove _init_model from ClusteringModel interface. We use abstract classes here as interfaces, so we can't write any method implementations inside. After removing _init_model you should remove __init__ of interface from GaussianMixtureClustering __init__ function

Reply:
In the case of the ClusteringModel interface, the _init_model method is meant to ensure that all clustering models know how to initialise their specific model. This is important because different clustering models may require different initialisation procedures.

In this case sklearn.mixture GaussianMixture initialises using self._init_model().

please read the following stack trace:

  File "/Users/Paul/git/Spaghetti/tests/clustering/test_gaussian_mixture_clustering.py", line 20, in setUp
    self.gmc = GaussianMixtureClustering()
  File "/Users/Paul/git/Spaghetti/spaghetti/clustering/GaussianMixtureClustering.py", line 33, in __init__
    super().__init__()
  File "/Users/Paul/git/Spaghetti/spaghetti/clustering/ClusteringModel.py", line 22, in __init__
    self._model = self._init_model()
AttributeError: 'GaussianMixtureClustering' object has no attribute '_init_model'

Reply:
I've suggested that we should not enforce the addition of private methods to the class. Yes, it is important to initiate the model, but it's not a must to do it with a separate private method. Another thing I was talking about is that we should not include any implementation into interfaces because that's the idea of the interface. I still think you should remove __init__ implementation from the ClusteringModel interface, therefore, remove super.__init__() from the  GaussianMixtureClustering __init__() method.